### PR TITLE
Added an allocation viewer

### PR DESCRIPTION
**Pull request by 1whatleytay**
This pull request adds an allocation viewer to the debug menu of Vita3K. The allocation viewer allows a debugger to do a few things:

- View the allocation's generation and name.
- Tell the size of the allocation in pages and kilobytes.
- View where allocation starts and ends (as a hexadecimal address).
- View and edit the contents of the allocation using imgui_club's memory editor.
screen shot 2019-01-10 at 10 24 41 pm
screen shot 2019-01-10 at 10 25 24 pm 2

### CODE REVIEW

File: src/emulator/gui/include/gui/state.h [hunk 1]
```diff
@@ -56,6 +58,8 @@ struct GuiState {
 
     DialogState common_dialog;
     GamesSelector game_selector;
+    MemoryEditor* memory_editor;
```
Comment (pent0) [inline src/emulator/gui/include/gui/state.h]:
```suggestion
    MemoryEditor *memory_editor;
```

File: src/emulator/gui/src/ui_allocations_dialog.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            host.gui.memory_editor->DrawWindow("Editor", host.mem.memory.get() + host.gui.memory_editor_start,
+                    (size_t) host.gui.memory_editor_count, (size_t) host.gui.memory_editor_start);
+        } else {
+            delete host.gui.memory_editor;
+            host.gui.memory_editor = nullptr;
+        }
+    }
+
+    ImGui::End();
+}
```
Comment (pent0) [inline src/emulator/gui/src/ui_allocations_dialog.cpp]: Newline

File: src/emulator/gui/src/ui_allocations_dialog.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <imgui.h>
+#include <imgui_memory_editor.h>
+
+#include <host/state.h>
+
+#include <string.h>
+
+void DrawAllocationsDialog(HostState &host) {
+    ImGui::Begin("Allocations", &host.gui.allocations_dialog);
```
Comment (VelocityRa) [inline src/emulator/gui/src/ui_allocations_dialog.cpp]:
```suggestion
    ImGui::Begin("Memory Allocations", &host.gui.allocations_dialog);
```

File: src/emulator/gui/src/ui_allocations_dialog.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    const std::lock_guard<std::mutex> lock(host.mem.generation_mutex);
+    for (const auto &pair : host.mem.generation_names) {
+        int length = snprintf(nullptr, 0, "%i: %s", (int)pair.first, pair.second.c_str());
```
Comment (VelocityRa) [inline src/emulator/gui/src/ui_allocations_dialog.cpp]:
```suggestion
        int length = snprintf(nullptr, 0, "%zu: %s", pair.first, pair.second.c_str());
```
For string formatting you could've also used [`fmt::format()`](http://fmtlib.net/latest/syntax.html), with the format string simply being `"{}: {}"`.

File: src/emulator/gui/src/ui_allocations_dialog.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        std::vector<char> buffer((unsigned long)length);
```
Comment (VelocityRa) [inline src/emulator/gui/src/ui_allocations_dialog.cpp]: Please avoid C-style casts, use C++ casts where it's necessary.

File: src/emulator/gui/src/ui_allocations_dialog.cpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        sprintf(&buffer[0], "%i: %s", (int)pair.first, pair.second.c_str());
```
Comment (VelocityRa) [inline src/emulator/gui/src/ui_allocations_dialog.cpp]: ditto about formatting

File: src/emulator/gui/src/ui_allocations_dialog.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                }
+            }
+            ImGui::TreePop();
+        }
+    }
+
+    if (host.gui.memory_editor) {
+        if (host.gui.memory_editor->Open) {
+            host.gui.memory_editor->DrawWindow("Editor", host.mem.memory.get() + host.gui.memory_editor_start,
+                    (size_t) host.gui.memory_editor_count, (size_t) host.gui.memory_editor_start);
```
Comment (VelocityRa) [inline src/emulator/gui/src/ui_allocations_dialog.cpp]: remove cast

File: src/emulator/gui/include/gui/state.h [hunk 8]
```diff
@@ -56,6 +58,8 @@ struct GuiState {
...
+    long memory_editor_start, memory_editor_count;
```
Comment (VelocityRa) [inline src/emulator/gui/include/gui/state.h]:
```suggestion
    size_t memory_editor_start, memory_editor_count;
```

File: src/emulator/gui/src/ui_main_menubar.cpp [hunk 9]
```diff
@@ -34,6 +34,7 @@ void DrawMainMenuBar(HostState &host) {
             ImGui::MenuItem("Condition Variables", nullptr, &host.gui.condvars_dialog);
             ImGui::MenuItem("Lightweight Condition Variables", nullptr, &host.gui.lwcondvars_dialog);
             ImGui::MenuItem("Event Flags", nullptr, &host.gui.eventflags_dialog);
+            ImGui::MenuItem("Allocations", nullptr, &host.gui.allocations_dialog);
```
Comment (VelocityRa) [inline src/emulator/gui/src/ui_main_menubar.cpp]:
```suggestion
            ImGui::MenuItem("Memory Allocations", nullptr, &host.gui.allocations_dialog);
```

**Comment by 1whatleytay** [comment]
Ahh, sorry for the test commit.

**Pull request merged**